File: stack/valid_paranthesis.py
# Q:20 - https://leetcode.com/problems/valid-parentheses/
class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        map = {")": "(", "]": "[", "}": "{"}

        for br in s:
            if br == "(" or br == "[" or br == "{":
                stack.append(br)
            elif br in map and stack and stack[-1] == map[br]:
                stack.pop()
            else:
                return False

        return len(stack) == 0



c = Solution()

print(c.isValid("()"))
print(c.isValid("()[]{}"))
print(c.isValid("(]"))

File: stack/valid_paranthesis.js
// Q:20 - https://leetcode.com/problems/valid-parentheses/

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = [];
  const map = new Map([
    [")", "("],
    ["]", "["],
    ["}", "{"],
  ]);

  for (let i = 0; i < s.length; i++) {
    const currentItem = s[i];
    if (!map.has(currentItem)) {
      stack.push(currentItem);
    } else {
      const lastItem = stack.pop();

      if (lastItem !== map.get(currentItem)) {
        return false;
      }
    }
  }

  return stack.length === 0;
};
